modules/server/src/main/scala/higherkindness/compendium/core/CompendiumService.scala

```diff
@@ -20,33 +20,44 @@ import cats.effect.Sync
 import cats.implicits._
 import higherkindness.compendium.core.refinements.ProtocolId
 import higherkindness.compendium.db.DBService
-import higherkindness.compendium.models.Protocol
+import higherkindness.compendium.models.DBModels.MetaProtocol
+import higherkindness.compendium.models.{IdlNames, Protocol, Target}
+import higherkindness.compendium.models.parserModels.ParserResult
+import higherkindness.compendium.parser.ProtocolParserService
 import higherkindness.compendium.storage.Storage
 
 trait CompendiumService[F[_]] {
 
-  def storeProtocol(id: ProtocolId, protocol: Protocol): F[Unit]
-  def recoverProtocol(protocolId: ProtocolId): F[Option[Protocol]]
+  def storeProtocol(id: ProtocolId, protocol: Protocol, idlName: IdlNames): F[Unit]
+  def recoverProtocol(protocolId: ProtocolId): F[Option[MetaProtocol]]
   def existsProtocol(protocolId: ProtocolId): F[Boolean]
+  def parseProtocol(protocolName: ProtocolId, target: Target): F[ParserResult]
 }
 
 object CompendiumService {
 
-  implicit def impl[F[_]: Sync: Storage: DBService: ProtocolUtils](): CompendiumService[F] =
+  implicit def impl[F[_]: Sync: Storage: DBService: ProtocolUtils: ProtocolParserService](): CompendiumService[
+    F] =
     new CompendiumService[F] {
 
-      override def storeProtocol(id: ProtocolId, protocol: Protocol): F[Unit] =
+      override def storeProtocol(id: ProtocolId, protocol: Protocol, idlName: IdlNames): F[Unit] =
         ProtocolUtils[F].validateProtocol(protocol) >>
-          DBService[F].upsertProtocol(id) >>
+          DBService[F].upsertProtocol(id, idlName) >>
           Storage[F].store(id, protocol)
 
-      override def recoverProtocol(protocolId: ProtocolId): F[Option[Protocol]] =
+      override def recoverProtocol(protocolId: ProtocolId): F[Option[MetaProtocol]] =
         DBService[F]
-          .existsProtocol(protocolId)
-          .ifM(Storage[F].recover(protocolId), none[Protocol].pure[F])
+          .selectProtocolById(protocolId)
```

Comment:
This would leave unhandled the error in case the protocol does not exist, returning `500` status error. I think `selectProtocolById` should return `Option` in case the protocol does not exist, so you do not need to check if it exists.